README.md
# JetEZ - Easy SDK

**BETA**

JetEZ is an easy (EZ) and lightweight alternative to the official JET SDK with some
limitations. JetEZ creates signed JunOS install packages including any kind of
files described in a [yaml file](https://git.juniper.net/cgiese/JetEZ/tree/master/docs/jet-yaml.md).
This allows to install Python and SLAX scripts but also cross compiled binaries.

The [example app](https://git.juniper.net/cgiese/JetEZ/tree/master/example-app)
shows how to build a GO application for JunOS using JetEZ.

In order to develop and distribute JetEZ applications, you need a package
[signing certificate](https://www.juniper.net/documentation/en_US/jet1.0/topics/task/jet-app-executing-certificate-request-script.html).

```
$ jetez --help
usage: jetez [-h] --source DIR [-v VERSION] -k FILE -c FILE [-j FILE]
             [-b BUILD] [-d]

JetEZ - Easy SDK
----------------

This tool creates a JET install package (.tgz) from source directory using
the parameters from jet.yaml project description file.

optional arguments:
  -h, --help            show this help message and exit
  --source DIR          source directory
  -v VERSION, --version VERSION
                        version string
  -k FILE, --key FILE   path to signing key
  -c FILE, --cert FILE  path to signing cert
  -j FILE, --jet FILE   path to project file (default: <source>/jet.yaml)
  -b BUILD, --build BUILD
                        build directory (default: .build)
  -d, --debug           verbose logging

Version: 0.3 Author: Christian Giese <cgiese@juniper.net>
```

JetEZ supports to generate signed install packages for OCCAM JunOS only (FreeBSD 10 or newer)!


CHANGELOG.md
# JetEZ - Release Notes

## 0.4

Add support for veriexec-ext, schema and config-validate. 

## 0.3

Add cert request script and support for files not linked (e.g. license, ...).

## 0.2

Extended support for project yaml file (jet.yaml) and python 2/3 compatibility.

## 0.1

First beta release supporting to generate signed JET install packages
for OCCAM JunOS (FreeBSD 10 or newer).


jet/utils.py
# Copyright 2018 Juniper Networks, Inc. All rights reserved
#
# Licensed under the Juniper Networks Script Software License (the "License").
# You may not use this script file except in compliance with the License, which is located at
# http://www.juniper.net/support/legal/scriptlicense/
# Unless required by applicable law or otherwise agreed to in writing by the parties,
# software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
# OR CONDITIONS OF ANY KIND, either express or implied.
"""JetEZ Library

:Organization: Juniper Networks, Inc.
:Copyright: Copyright (c) 2018, Juniper Networks, Inc. All rights reserved.
:Author: Christian Giese
:Contact: cgiese@juniper.net
:Date: 05/23/2018
:Version: 0.4
"""
from contextlib import contextmanager
from jet import crypto
from lxml import etree
from yaml.error import YAMLError
import argparse
import binascii
import datetime
import logging
import os
import subprocess
import shutil
import yaml
import sys

try:
    FileNotFoundError
except NameError:
    # python 2 compatibility
    FileNotFoundError = IOError

log = logging.getLogger("jet")


@contextmanager
def exit_on_error(*exceptions):
    """log exception and exit

    :param exceptions: exceptions
    :type exceptions: Exception
    """
    try:
        yield
    except YAMLError as error:
        log.error("failed to parse yaml file\n %s", error)
        sys.exit(1)
    except FileNotFoundError as error:
        log.error("no such file or directory: %s", error.filename)
        sys.exit(1)
    except Exception as error:
        log.error(str(error))
        sys.exit(1)


class FileArgumentParser(argparse.ArgumentParser):

    def __file_error(self, metavar, arg, args):
        self.error("argument %s: %s %s not found" % ("/".join(args), metavar, arg))

    def __is_valid_file(self, arg, args):
        if not os.path.isfile(arg):
            self.__file_error("file", arg, args)
        else:
            # File exists so return the filename
            return arg

    def __is_valid_directory(self, arg, args):
        if not os.path.isdir(arg):
            self.__file_error("directory", arg, args)
        else:
            # File exists so return the directory
            return arg

    def add_argument_with_check(self, *args, **kwargs):
        # Look for your FILE or DIR settings
        if 'metavar' in kwargs and 'type' not in kwargs:
            if kwargs['metavar'] is 'FILE':
                type=lambda x: self.__is_valid_file(x, args)
                kwargs['type'] = type
            elif kwargs['metavar'] is 'DIR':
                type=lambda x: self.__is_valid_directory(x, args)
                kwargs['type'] = type
        self.add_argument(*args, **kwargs)


def load_project(project_file, version):
    """load jet project from yaml file

    :param project_file: path to input file
    :type project_file: str
    :return: jet project
    :rtype: dict
    """
    with open(project_file , 'r') as f:
        project_yaml = yaml.load(f)

    def required(d, k):
        if k in d:
            return d[k]
        else:
            raise ValueError("missing attribute %s" % k)

    timestamp = datetime.datetime.now()
    project = {
        "basename": required(project_yaml, "basename"),
        "files": [],
        "comment": project_yaml.get("comment", "JET app %s" % project_yaml["basename"]),
        "arch": project_yaml.get("arch", "x86"),
        "abi": project_yaml.get("abi", "32"),
        "copyright": project_yaml.get("copyright", "Copyright %s, Juniper Networks, Inc." % timestamp.year),
        "package_id": project_yaml.get("package_id", 31),
        "role": project_yaml.get("role", "Provider_Daemon"),
        "date": timestamp.strftime("%Y%m%d"),
        "time": timestamp.strftime("%H%M%S"),
        "schema": project_yaml.get("schema", False),
        "config-validate": project_yaml.get("config-validate", False),
        "veriexec-ext": project_yaml.get("veriexec-ext", False),
    }
    project_yaml_files = required(project_yaml, "files")
    for file in project_yaml_files:
        project["files"].append({
            "source": required(file, "source"),
            "destination": required(file, "destination"),
            "uid": file.get("uid", 0),
            "gid": file.get("gid", 0),
            "mode": file.get("mode", 555),
            "program_id": file.get("program_id", 1),
            "symlink": file.get("symlink", True),
        })
    return project


def create_contents_iso(input, output):
    """create iso file

    :param input: path to input file
    :type input: str
    :param output: path to output file
    :type output: str
    """
    subprocess.check_output("mkisofs --rock -sysid JUNOS -o %s %s" % (output, input), shell=True)


def create_tgz(package, path):
    """create tgz archive

    :param package: package name
    :type package: str
    :param path: path to build directory
    :type path: str
    """
    files = " ".join(os.listdir(path))
    subprocess.check_output("tar -C %s -czf %s.tgz %s" % (path, package, files), shell=True)


def create_package_xml(project, version, package, path):
    """create package.xml file

    :param project: jet project
    :type project: dict
    :param version: version string
    :type version: str
    :param package: package name
    :type package: str
    :param path: path to build directory
    :type path: str
    """
    def package_xml_file(filename):
        filename = os.path.join(path, filename)
        _file = etree.SubElement(dir, "file", name=os.path.basename(filename))
        etree.SubElement(_file, "size").text = str(os.path.getsize(filename))
        etree.SubElement(_file, "sha256").text = str(crypto.generate_sha256(filename))

    package_xml = etree.Element("package", name=package)
    etree.SubElement(package_xml, "abi").text = project["abi"]
    etree.SubElement(package_xml, "arch").text = project["arch"]
    etree.SubElement(package_xml, "basename").text = project["basename"]
    etree.SubElement(package_xml, "comment").text = "%s [%s]" % (project["comment"], version)
    etree.SubElement(package_xml, "copyright").text = project["copyright"]
    etree.SubElement(package_xml, "description").text = project["basename"]
    etree.SubElement(package_xml, "mntname").text = "%s%s-%s" % (project["basename"], project["abi"], binascii.b2a_hex(os.urandom(4)).decode())
    etree.SubElement(package_xml, "require").text = "junos-runtime32"
    etree.SubElement(package_xml, "version").text = project["date"]
    etree.SubElement(package_xml, "spin").text = project["time"]
    etree.SubElement(package_xml, "sb-location").text = "JetEZ"

    # XMLPKG_TOGGLE_LIST
    for p in ("schema", "config-validate", "veriexec-ext"):
        if project[p]:
            etree.SubElement(package_xml, p)

    dir = etree.SubElement(package_xml, "dir", name="contents")
    package_xml_file("contents/contents.iso")
    package_xml_file("contents/contents.symlinks")
    with open("%s/package.xml" % path, "w+") as f:
        f.write(etree.tostring(package_xml, pretty_print=True).decode("utf8"))


jet/crypto.py
# Copyright 2018 Juniper Networks, Inc. All rights reserved
#
# Licensed under the Juniper Networks Script Software License (the "License").
# You may not use this script file except in compliance with the License, which is located at
# http://www.juniper.net/support/legal/scriptlicense/
# Unless required by applicable law or otherwise agreed to in writing by the parties,
# software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
# OR CONDITIONS OF ANY KIND, either express or implied.
"""JetEZ Library

:Organization: Juniper Networks, Inc.
:Copyright: Copyright (c) 2018, Juniper Networks, Inc. All rights reserved.
:Author: Christian Giese
:Contact: cgiese@juniper.net
:Date: 04/12/2018
:Version: 0.4
"""
import base64
import logging
import hashlib
import os
import subprocess

FILE_BUF_SIZE=65536

log = logging.getLogger("jet")


def generate_sha256(filename):
    """generate sha256 hash of file

    :param filename: filename
    :type filename: str
    :return: sha256 hash of file
    :rtype: str
    """
    hash = hashlib.sha256()
    with open(filename, 'rb') as f:
        while True:
            data = f.read(FILE_BUF_SIZE)
            if not data:
                break
            hash.update(data)
    return hash.hexdigest()


def generate_sha1(filename):
    """generate sha1 hash of file

    :param filename: filename
    :type filename: str
    :return: sha1 hash of file
    :rtype: str
    """
    hash = hashlib.sha1()
    with open(filename, 'rb') as f:
        while True:
            data = f.read(FILE_BUF_SIZE)
            if not data:
                break
            hash.update(data)
    return hash.hexdigest()


def sign(input, output, key, cert):
    """sign file

    :param input: path to input file
    :type input: str
    :param output: path to output file
    :type output: str
    :param key: path to private key
    :type key: str
    :param cert: path to public key/cert
    :type cert: str
    """
    # get subject from certificate
    _subject = subprocess.check_output("openssl x509 -in %s -noout -subject" % cert, shell=True)
    subject = _subject.decode("utf8").split(" ", 1)[1].strip()
    # create signature
    signature = subprocess.check_output("openssl dgst -sha1 -sign %s %s" % (key, input), shell=True)
    # create base64 from signature
    signature64 = base64.b64encode(signature).decode("utf8")
    # format signature file
    signature_file="""%s
-----BEGIN JUNOS SIGNATURE-----
%s
-----END JUNOS SIGNATURE-----
""" % (subject, "\n".join([signature64[x:x+64] for x in range(0, len(signature64), 64)]))
    # write signature file
    with open(output, "w+") as f:
        f.write(signature_file)
    # create certificate chain
    with open("%s/manifest.certs" % os.path.dirname(__file__), "r") as f:
        cert_chain = f.read()
    with open(cert, "r") as f:
        cert_file = f.read()
    with open("%s/manifest.certs" % os.path.dirname(output), "w+") as f:
        f.write(cert_file)
        f.write(cert_chain)


cert-request/ask.sh
:
# RCSid:
#	$Id: ask.sh 493955 2012-01-31 04:15:37Z ib-builder $
#
#	@(#) Copyright (c) 1994 Simon J. Gerraty
#
#	This file is provided in the hope that it will
#	be of use.  There is absolutely NO WARRANTY.
#	Permission to copy, redistribute or otherwise
#	use this file is hereby granted provided that 
#	the above copyright notice and this notice are
#	left intact. 
#      
#	Please send copies of changes and bug-fixes to:
#	sjg@crufty.net
#

case `echo -n .` in -n*) N=; C="\c";; *) N="-n"; C=;; esac

_TTY=${_TTY:-`test -t 0 && tty`}; export _TTY

_ask() {
	if test "$_TTY"; then
		echo $N "${2:-$1?} "$C > $_TTY
		read $1 < $_TTY
	fi
}

_ask_noecho() {
	if test "$_TTY"; then
		stty -echo < $_TTY
		echo $N "${2:-$1?} "$C > $_TTY
		read $1 < $_TTY
		echo > $_TTY
		stty echo < $_TTY
	fi
}

# it would be nice to use local var etc, but not all shells support it.
ask() {
	case "$1" in
	-echo)	_askfunc=_ask_noecho; shift;;
	*)	_askfunc=_ask;;
	esac
	$_askfunc _a "$2${3:+ [}$3${3:+]}:"
	eval "$1='${_a:-$3}'"
}

# This form is suitable for foo=`asks prompt`
asks() {
	case "$1" in
	-echo)	_askfunc=_ask_noecho; shift;;
	*)	_askfunc=_ask;;
	esac
	$_askfunc __a "$@"
	echo $__a
}

no_help() {
	echo "Sorry, no help available for $1"
}

askv() {
	case "$1" in
	-echo)	_askfunc=_ask_noecho; shift;;
	*)	_askfunc=_ask;;
	esac
	while :
	do
		test "$_TTY" || break
		eval "$_askfunc _a \"$2${3:+ [}$3${3:+]}:\""
		
		case "$_a" in
		\?)	eval \${help_$1:-${help_func:-no_help}} $1
			continue
			;;
		esac
		eval \${validate_$1:-${validate_func:-:}} $1 "${_a:-$3}" && break
	done
	eval "$1='${_a:-$3}'"
}

case $0 in
*asks*)	asks "$@";;
*askv*)	askv "$@"; eval echo $1=\$$1;;
*ask*)	ask "$@"; eval echo $1=\$$1;;
esac

       


jet/main.py
#!/usr/bin/env python
#
# Copyright 2018 Juniper Networks, Inc. All rights reserved
#
# Licensed under the Juniper Networks Script Software License (the "License").
# You may not use this script file except in compliance with the License, which is located at
# http://www.juniper.net/support/legal/scriptlicense/
# Unless required by applicable law or otherwise agreed to in writing by the parties,
# software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
# OR CONDITIONS OF ANY KIND, either express or implied.
"""JetEZ

:Organization: Juniper Networks, Inc.
:Copyright: Copyright (c) 2018, Juniper Networks, Inc. All rights reserved.
:Author: Christian Giese
:Contact: cgiese@juniper.net
:Date: 04/19/2018
:Version: 0.4
"""
from jet import crypto
from jet import utils
import argparse
import datetime
import logging
import shutil
import sys
import os
import yaml

DESCIPTION = """
JetEZ - Easy SDK
----------------

This tool creates a JET install package (.tgz) from source directory using
the parameters from jet.yaml project description file.

"""

# setup logging to stdout
log = logging.getLogger("jet")
log.setLevel(logging.DEBUG)
ch = logging.StreamHandler(sys.stdout)
ch.setLevel(logging.INFO)
ch.setFormatter(logging.Formatter('%(levelname)s: %(message)s'))
log.addHandler(ch)

def main():
    parser = utils.FileArgumentParser(description=DESCIPTION, formatter_class=argparse.RawDescriptionHelpFormatter,
                                      epilog="Version: 0.3 Author: Christian Giese <cgiese@juniper.net>")
    parser.add_argument_with_check('--source', dest='source', metavar='DIR', required=True, help="source directory")
    parser.add_argument('-v', '--version', dest='version', type=str, help="version string")
    parser.add_argument_with_check('-k', '--key', dest='key', metavar='FILE', required=True, help="path to signing key")
    parser.add_argument_with_check('-c', '--cert', dest='cert', metavar='FILE', required=True, help="path to signing cert")
    parser.add_argument_with_check('-j', '--jet', dest='jet', metavar='FILE', required=False, help="path to project file (default: <source>/jet.yaml)")
    parser.add_argument('-b', '--build', dest='build', default=".build", help="build directory (default: .build)")
    parser.add_argument('-d', '--debug', dest='debug', action='store_true', help="verbose logging")
    args = parser.parse_args()
    if args.debug:
        ch.setLevel(logging.DEBUG)

    version = args.version if args.version else datetime.datetime.now().strftime("%Y%m%d-%H%M%S")

    project_file = args.jet if args.jet else os.path.join(args.source, "jet.yaml")
    log.info("load project file %s", project_file)
    with utils.exit_on_error(Exception):
        project = utils.load_project(project_file, version)
    package = "%s-%s-%s-%s" % (project["basename"], project["arch"], project["abi"], version)

    log.info("create temporary build directory %s", args.build)
    if os.path.exists(args.build):
        shutil.rmtree(args.build)
    os.makedirs(args.build)
    os.makedirs('%s/contents' % args.build)
    contents = "%s/contents/contents" % args.build
    os.makedirs(contents)
    contents_pkg = '%s/pkg' % contents
    os.makedirs(contents_pkg)
    content_manifest = """pkg/manifest uid=0 gid=0 mode=444
pkg/manifest.sha1 uid=0 gid=0 mode=444
pkg/manifest.sig uid=0 gid=0 mode=444
pkg/manifest.certs uid=0 gid=0 mode=444
/set package_id=%s role=%s
""" % (project["package_id"], project["role"])
    contents_symlink =""
    mount_dir = "/packages/mnt/%s" % project["basename"]
    for f in project["files"]:
        destination = os.path.join(contents, *f['destination'].split("/"))
        log.info("copy file %s to %s", f['source'], destination)
        # create target directory
        _d = f['destination'].split("/")[1:]
        for d in  _d[:1] if len(_d) == 2 else ["/".join(_d[:x]) for x in range(2, len(_d))]:
            _d = os.path.join(contents, d)
            if not os.path.exists(_d):
                os.makedirs(_d)
        # copy file
        shutil.copy(os.path.join(args.source, f['source']), destination)
        # add file to manifest
        sha1 = crypto.generate_sha1(destination)
        content_manifest += "%s sha1=%s uid=%s gid=%s mode=%s program_id=%s\n" % \
            (f["destination"][1:] if f["destination"][0] == "/" else f["destination"],
             sha1, f["uid"], f["gid"], f["mode"], f["program_id"])
        if f["symlink"]:
            contents_symlink += "%s%s %s\n" % (mount_dir, f["destination"], f["destination"])

    content_manifest_file = '%s/manifest' % contents_pkg
    log.info("create manifest file %s", content_manifest_file)
    with open(content_manifest_file, "w") as f:
        f.write(content_manifest)

    content_manifest_sha_file = '%s/manifest.sha1' % contents_pkg
    with open(content_manifest_sha_file, "w") as f:
        f.write("%s\n" % crypto.generate_sha1(content_manifest_file))

    contents_symlink_file = '%s.symlinks' % contents
    log.info("create symlink file %s", contents_symlink_file)
    with open(contents_symlink_file, "w") as f:
        f.write(contents_symlink)

    log.info("sign manifest file %s" % content_manifest_file)
    crypto.sign(content_manifest_file, "%s.sig" % content_manifest_file, args.key, args.cert)

    for f in os.listdir(contents_pkg):
        os.chmod(os.path.join(contents_pkg, f), 0o444)

    log.info("create contents.iso")
    utils.create_contents_iso(contents, "%s.iso" % contents)
    shutil.rmtree(contents)

    log.info("create package.xml")
    utils.create_package_xml(project, args.version, package, args.build)

    package_manifest = "/set package_id=31 role=Provider_Daemon\n"
    package_manifest_files = ["contents/contents.iso", "contents/contents.symlinks", "package.xml"]

    for f in package_manifest_files:
        package_manifest += "%s sha1=%s\n" % (f, crypto.generate_sha1(os.path.join(args.build, f)))

    package_manifest_file = os.path.join(args.build, "manifest")
    log.info("create manifest file %s", package_manifest_file)
    with open(package_manifest_file, "w") as f:
        f.write(package_manifest)

    log.info("sign manifest file %s" % package_manifest_file)
    crypto.sign(package_manifest_file, "%s.sig" % package_manifest_file, args.key, args.cert)

    log.info("create %s.tgz" % package)
    utils.create_tgz(package, args.build)

    log.info("package successfully created")

if __name__ == "__main__":
    main()
